core/src/avm2/globals/flash/geom/Vector3D.as

```diff
@@ -0,0 +1,146 @@
+package flash.geom {
+    public class Vector3D {
+
+        // TODO: does the order matter?
+        public var w:Number;
+        public var x:Number;
+        public var y:Number;
+        public var z:Number;
+
+        // TODO: and here?
+        public static const X_AXIS : Vector3D = new Vector3D(1, 0, 0);
+        public static const Y_AXIS : Vector3D = new Vector3D(0, 1, 0);
+        public static const Z_AXIS : Vector3D = new Vector3D(0, 0, 1);
+
+        public function Vector3D(x:Number = 0, y:Number = 0, z:Number = 0, w:Number = 0) {
+            this.x = x;
+            this.y = y;
+            this.z = z;
+            this.w = w;
+        }
+
+        public function get lengthSquared():Number {
+            return this.x * this.x + this.y * this.y + this.z * this.z;
+        }
+
+        public function get length():Number {
+            return Math.sqrt(this.lengthSquared);
+        }
+
+        public function toString():String {
+            return "Vector3D(" + this.x + ", " + this.y + ", " + this.z + ")";
+        }
+
+        public function add(a:Vector3D):Vector3D {
+            return new Vector3D(this.x + a.x, this.y + a.y, this.z + a.z);
+        }
+
+        public function subtract(a:Vector3D):Vector3D {
+            return new Vector3D(this.x - a.x, this.y - a.y, this.z - a.z);
+        }
+
+        public function incrementBy(a:Vector3D):void {
+            this.x += a.x;
+            this.y += a.y;
+            this.z += a.z;
+        }
+
+        public function decrementBy(a:Vector3D):void {
+            this.x -= a.x;
+            this.y -= a.y;
+            this.z -= a.z;
+        }
+
+        public function clone():Vector3D {
+            return new Vector3D(this.x, this.y, this.z, this.w);
+        }
+
+        public function copyFrom(sourceVector3D:Vector3D):void {
+            this.x = sourceVector3D.x;
+            this.y = sourceVector3D.y;
+            this.z = sourceVector3D.z;
+            // w is not touched!
+        }
+
+        public function equals(toCompare:Vector3D, allFour:Boolean = false):Boolean {
+            return this.x == toCompare.x && this.y == toCompare.y && this.z == toCompare.z && (!allFour || this.w == toCompare.w);
+        }
+
+        public function nearEquals(toCompare:Vector3D, tolerance:Number, allFour:Boolean = false):Boolean {
+            if (allFour)
+                // TODO seems like in this case, tolerance is just ignored???
+                return this.equals(toCompare, true);
+
+            return (Math.abs(this.x - toCompare.x) < tolerance)
+                && (Math.abs(this.y - toCompare.y) < tolerance)
+                && (Math.abs(this.z - toCompare.z) < tolerance);
+        }
+
+        public function setTo(xa:Number, ya:Number, za: Number):void {
+            this.x = xa;
+            this.y = ya;
+            this.z = za;
```

Comment:
I would like you to add the comment `// w is unchanged!` here and where else it's relevant.

Reply:
Where I did not write that, I did so because the AS3 reference mentioned that only x, y, and z are affected. Where I put the comments, I had to figure this out. But you are right, just looking at the code itself, it helps if it's everywhere.

Reply:
I added some comments, if you'd like me to add them to some more places, do tell.